### PR TITLE
Port to Python3

### PR DESCRIPTION
**Pull request by srevinsaju**
I need assistance for testing the port. Please help until i get my environment ready

### CODE REVIEW

**Comment by quozl** [comment]
Needs Port to TelepathyGLib first.  Ubuntu and Fedora do not have a Python 3 build of Telepathy using "import telepathy".

**Comment by srevinsaju** [comment]
`pip3 install telepathy --user`

**Comment by chimosky** [comment]
> `pip3 install telepathy --user`

The activity still needs to be ported to TelepathyGLib, look at the porting guide.

**Comment by srevinsaju** [comment]
#10 @chimosky @quozl

**Comment by chimosky** [comment]
Nothing's changed from the last commit you made and the activity needs to be ported to TelepathyGLib.

**Comment by chimosky** [comment]
I've looked at 5b340fb and it's the same as what you did in #10, check to make sure you understand the comments in #10 else let us know and we'll explain further.

**Comment by srevinsaju** [comment]
@chimosky I have edited the iBus channels too. The previous commit didn't have that
Apart from using import ... as telepathy, the rest everything comprehends to GLibTelepathy.

**Comment by srevinsaju** [comment]
I have did what the porting guide told me to do, but it seems that the ibus cannot connect. Is it an issue?
I can run colordeducto, but see what happens when you press the gear icon.

```
(sugar-activity3:4922): IBUS-WARNING **: 16:02:44.850: Unable to connect to ibus: Could not connect: Connection refused
Traceback (most recent call last):
  File "/usr/lib64/python3.7/site-packages/dbus/bus.py", line 175, in activate_name_owner
    return self.get_name_owner(bus_name)
  File "/usr/lib64/python3.7/site-packages/dbus/bus.py", line 361, in get_name_owner
    's', (bus_name,), **keywords)
  File "/usr/lib64/python3.7/site-packages/dbus/connection.py", line 651, in call_blocking
    message, timeout)
dbus.exceptions.DBusException: org.freedesktop.DBus.Error.NameHasNoOwner: The name does not have an owner

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/srevinsaju/repo/colordeducto/ColorDeductoActivity.py", line 281, in _gear_cb
    self._load_python_code_from_journal)
  File "/home/srevinsaju/repo/colordeducto/ColorDeductoActivity.py", line 317, in _chooser
    result = chooser.run()
  File "/usr/local/lib/python3.7/site-packages/sugar3/graphics/objectchooser.py", line 214, in run
    obj = self._bus.get_object(J_DBUS_SERVICE, J_DBUS_PATH)
  File "/usr/lib64/python3.7/site-packages/dbus/bus.py", line 241, in get_object
    follow_name_owner_changes=follow_name_owner_changes)
  File "/usr/lib64/python3.7/site-packages/dbus/proxies.py", line 248, in __init__
    self._named_service = conn.activate_name_owner(bus_name)
  File "/usr/lib64/python3.7/site-packages/dbus/bus.py", line 180, in activate_name_owner
    self.start_service_by_name(bus_name)
  File "/usr/lib64/python3.7/site-packages/dbus/bus.py", line 278, in start_service_by_name
    'su', (bus_name, flags)))
  File "/usr/lib64/python3.7/site-packages/dbus/connection.py", line 651, in call_blocking
    message, timeout)
dbus.exceptions.DBusException: org.freedesktop.DBus.Error.ServiceUnknown: The name is not activatable
Traceback (most recent call last):
  File "/usr/lib64/python3.7/site-packages/dbus/bus.py", line 175, in activate_name_owner
    return self.get_name_owner(bus_name)
  File "/usr/lib64/python3.7/site-packages/dbus/bus.py", line 361, in get_name_owner
    's', (bus_name,), **keywords)
  File "/usr/lib64/python3.7/site-packages/dbus/connection.py", line 651, in call_blocking
    message, timeout)
dbus.exceptions.DBusException: org.freedesktop.DBus.Error.NameHasNoOwner: The name does not have an owner

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/srevinsaju/repo/colordeducto/ColorDeductoActivity.py", line 281, in _gear_cb
    self._load_python_code_from_journal)
  File "/home/srevinsaju/repo/colordeducto/ColorDeductoActivity.py", line 317, in _chooser
    result = chooser.run()
  File "/usr/local/lib/python3.7/site-packages/sugar3/graphics/objectchooser.py", line 214, in run
    obj = self._bus.get_object(J_DBUS_SERVICE, J_DBUS_PATH)
  File "/usr/lib64/python3.7/site-packages/dbus/bus.py", line 241, in get_object
    follow_name_owner_changes=follow_name_owner_changes)
  File "/usr/lib64/python3.7/site-packages/dbus/proxies.py", line 248, in __init__
    self._named_service = conn.activate_name_owner(bus_name)
  File "/usr/lib64/python3.7/site-packages/dbus/bus.py", line 180, in activate_name_owner
    self.start_service_by_name(bus_name)
  File "/usr/lib64/python3.7/site-packages/dbus/bus.py", line 278, in start_service_by_name
    'su', (bus_name, flags)))
  File "/usr/lib64/python3.7/site-packages/dbus/connection.py", line 651, in call_blocking
    message, timeout)
dbus.exceptions.DBusException: org.freedesktop.DBus.Error.ServiceUnknown: The name is not activatable
Traceback (most recent call last):
  File "/usr/lib64/python3.7/site-packages/dbus/bus.py", line 175, in activate_name_owner
    return self.get_name_owner(bus_name)
  File "/usr/lib64/python3.7/site-packages/dbus/bus.py", line 361, in get_name_owner
    's', (bus_name,), **keywords)
  File "/usr/lib64/python3.7/site-packages/dbus/connection.py", line 651, in call_blocking
    message, timeout)
dbus.exceptions.DBusException: org.freedesktop.DBus.Error.NameHasNoOwner: The name does not have an owner

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/srevinsaju/repo/colordeducto/ColorDeductoActivity.py", line 281, in _gear_cb
    self._load_python_code_from_journal)
  File "/home/srevinsaju/repo/colordeducto/ColorDeductoActivity.py", line 317, in _chooser
    result = chooser.run()
  File "/usr/local/lib/python3.7/site-packages/sugar3/graphics/objectchooser.py", line 214, in run
    obj = self._bus.get_object(J_DBUS_SERVICE, J_DBUS_PATH)
  File "/usr/lib64/python3.7/site-packages/dbus/bus.py", line 241, in get_object
    follow_name_owner_changes=follow_name_owner_changes)
  File "/usr/lib64/python3.7/site-packages/dbus/proxies.py", line 248, in __init__
    self._named_service = conn.activate_name_owner(bus_name)
  File "/usr/lib64/python3.7/site-packages/dbus/bus.py", line 180, in activate_name_owner
    self.start_service_by_name(bus_name)
  File "/usr/lib64/python3.7/site-packages/dbus/bus.py", line 278, in start_service_by_name
    'su', (bus_name, flags)))
  File "/usr/lib64/python3.7/site-packages/dbus/connection.py", line 651, in call_blocking
    message, timeout)
dbus.exceptions.DBusException: org.freedesktop.DBus.Error.ServiceUnknown: The name is not activatable
Traceback (most recent call last):
  File "/usr/lib64/python3.7/site-packages/dbus/bus.py", line 175, in activate_name_owner
    return self.get_name_owner(bus_name)
  File "/usr/lib64/python3.7/site-packages/dbus/bus.py", line 361, in get_name_owner
    's', (bus_name,), **keywords)
  File "/usr/lib64/python3.7/site-packages/dbus/connection.py", line 651, in call_blocking
    message, timeout)
dbus.exceptions.DBusException: org.freedesktop.DBus.Error.NameHasNoOwner: The name does not have an owner

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/srevinsaju/repo/colordeducto/ColorDeductoActivity.py", line 281, in _gear_cb
    self._load_python_code_from_journal)
  File "/home/srevinsaju/repo/colordeducto/ColorDeductoActivity.py", line 317, in _chooser
    result = chooser.run()
  File "/usr/local/lib/python3.7/site-packages/sugar3/graphics/objectchooser.py", line 214, in run
    obj = self._bus.get_object(J_DBUS_SERVICE, J_DBUS_PATH)
  File "/usr/lib64/python3.7/site-packages/dbus/bus.py", line 241, in get_object
    follow_name_owner_changes=follow_name_owner_changes)
  File "/usr/lib64/python3.7/site-packages/dbus/proxies.py", line 248, in __init__
    self._named_service = conn.activate_name_owner(bus_name)
  File "/usr/lib64/python3.7/site-packages/dbus/bus.py", line 180, in activate_name_owner
    self.start_service_by_name(bus_name)
  File "/usr/lib64/python3.7/site-packages/dbus/bus.py", line 278, in start_service_by_name
    'su', (bus_name, flags)))
  File "/usr/lib64/python3.7/site-packages/dbus/connection.py", line 651, in call_blocking
    message, timeout)
dbus.exceptions.DBusException: org.freedesktop.DBus.Error.ServiceUnknown: The name is not activatable
[srevinsaju@localhost colordeducto]$ 
```

These errors have no references to the internal file, making me more confused. Your assistance is appreciated :)

**Comment by Hrishi1999** [comment]
(Discussion from #sugar) This error is caused because of you trying to use `sugar-activity .` to test the activity. Please refrain from using it :)

File: ColorDeductoActivity.py [hunk 1]
```diff
@@ -389,8 +389,8 @@ def _new_tube_cb(self, id, initiator, type, service, params, state):
 params={} state={}'.format(id, initiator, type, service, params, state))
 
         if (type == telepathy.IFACE_CHANNEL_TYPE_DBUS_TUBE and service == SERVICE):
-            # FIXME if state == telepathy.TUBE_STATE_LOCAL_PENDING: was changed as follows
-            if state == telepathy.PROP_CHANNEL_INTERFACE_GROUP_LOCAL_PENDING_MEMBERS:
+            
+            if state == TelepathyGLib.TubeState(0):
```
Comment (chimosky) [inline ColorDeductoActivity.py]: We normally use `TelepathyGLib.TubeState.LOCAL_PENDING`.

Comment (chimosky) [reply ColorDeductoActivity.py]: Seen your change in fd45a89

**Comment by chimosky** [comment]
Errors are expected as your change isn't right. See this [commit](https://github.com/sugarlabs/collabwrapper/commit/a6f9cc2dc4435841cdf3dc6d96635553d12ac2c4) for an idea on how to port to TelepathyGLib.

**Comment by srevinsaju** [comment]
The  commit https://github.com/sugarlabs/colordeducto/pull/9/commits/fd45a89fc99ab1241e2c8d4cd7a1d49b8295e989 fixes the code for Telepathy port and Python3 port.  Special thanks to @Hrishi1999 for helping me out with errors I faced, and methods on how to solve them. :100: 

@chimosky I referred to collabwrapper.py as by port guide. Please review and thanks

**Comment by chimosky** [comment]
Seems you haven't pushed your recent changes.

**Comment by srevinsaju** [comment]
@chimosky I have pushed everything, is there anything I have to change ?

**Comment by quozl** [comment]
Thanks for the progress.  Code review of the [aggregate change](https://github.com/sugarlabs/colordeducto/pull/9/files).

* [ ] please don't rename TelepathyGLib as telepathy during port; (a) because we don't do that, and (b) it makes searches for yet to be ported activities difficult,
* [ ] please don't rename constants you don't then go on to use, e.g. CHANNEL,
* [ ] please don't mix the renaming practice with direct access; i.e. if you are going to use renaming of the constants back to telepathy form, then all constants should be treated the same.
* [ ] please don't leave removed lines as commented code, it misleads new readers.

**Comment by srevinsaju** [comment]
> Thanks for the progress. Code review of the [aggregate change](https://github.com/sugarlabs/colordeducto/pull/9/files).
> 
> * [x]  please don't rename TelepathyGLib as telepathy during port; (a) because we don't do that, and (b) it makes searches for yet to be ported activities difficult,
> * [x]  please don't rename constants you don't then go on to use, e.g. CHANNEL,
> * [x]  please don't mix the renaming practice with direct access; i.e. if you are going to use renaming of the constants back to telepathy form, then all constants should be treated the same.
> * [x]  please don't leave removed lines as commented code, it misleads new readers.
I have done these in commits https://github.com/sugarlabs/colordeducto/pull/9/commits/8e6ab5fae023c59317676f431e61a284f30f09a2 and https://github.com/sugarlabs/colordeducto/pull/9/commits/6d075accf214c31f12f3c80ef312d5d0257fb8ce

**Comment by quozl** [comment]
Thanks.  Did a collaboration test of 6d075ac.  Sharer cannot share; when shared, the instance does not appear on other network views, and logs of sharer contain;

```log
Traceback (most recent call last):
  File "/usr/share/sugar/activities/ColorDeducto.activity/ColorDeductoActivity.py", line 328, in _shared_cb
    self._new_tube_common(True)
  File "/usr/share/sugar/activities/ColorDeducto.activity/ColorDeductoActivity.py", line 336, in _new_tube_common
    if self._shared_activity is None:
AttributeError: 'ColorDeductoActivity' object has no attribute '_shared_activity'
```

This is a canary to show no collaboration testing since 6bef019f4ab9c008109e5a06a0530cc277d2bc0c in 2012.  Fixed in 8e19973 pushed here.

Tested collaboration again.  Sharer does share.  Joiner can join.  However the activity doesn't seem to share any game state and doesn't log any errors.  This would appear to be a problem that was in the activity before you began; [making sure the activity works properly before porting](https://github.com/sugarlabs/sugar-docs/blob/master/src/python-porting-guide.md#how-to-port-an-activity-to-python-3) was in the guide, but you must have missed that step. :grin:

**Comment by srevinsaju** [comment]
phew, I didn't check if it works before porting. there are some more code formatting left to be done, which I will push later, @Quozl, thanks for testing :)

**Pull request merged**